paasta_tools/autoscaling_lib.py

```diff
@@ -569,104 +589,165 @@ def sort_slaves_to_kill(mesos_state, pool='default'):
-    ec2_client = boto3.client('ec2')
     if delta == 0:
         return
     elif delta > 0:
         log.info("Increasing spot fleet capacity to: {0}".format(target_capacity))
-        if not dry_run:
-            ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
-                                                 ExcessCapacityTerminationPolicy='noTermination')
-            return
+        set_spot_fleet_request_capacity(sfr_id, target_capacity, dry_run)
+        return
     elif delta < 0:
-        number_to_kill = delta * -1
-        sfr_ips = get_sfr_instance_ips(sfr_id)
-        log.debug("IPs in SFR: {0}".format(sfr_ips))
-        sfr_sorted_slaves = [slave for slave in sorted_slaves if slave_pid_to_ip(slave['pid']) in sfr_ips]
+        sfr_sorted_slaves = filter_sfr_slaves(sorted_slaves, resource['sfr'])
         log.info("SFR slave kill preference: {0}".format([slave['pid'] for slave in sfr_sorted_slaves]))
-        if number_to_kill > len(sfr_sorted_slaves):
+        killable_capacity = sum([slave['instance_weight'] for slave in sfr_sorted_slaves])
+        amount_to_decrease = delta * -1
+        if amount_to_decrease > killable_capacity:
             log.error("Didn't find enough candidates to kill. This shouldn't happen so let's not kill anything!")
             return
-        slaves_to_kill = sfr_sorted_slaves[0:number_to_kill]
-        log.info("Set to kill: {0}".format([slave['pid'] for slave in slaves_to_kill]))
-        instances_to_kill = {}
-        for slave in slaves_to_kill:
-            ip = slave_pid_to_ip(slave['pid'])
-            instances_to_kill[slave['pid']] = {'ip': ip,
-                                               'instance_id': get_instance_id_from_ip(ip)}
-        # The start time of the maintenance window is the point at which
-        # we giveup waiting for the instance to drain and mark it for termination anyway
-        start = int(time.time() + CLUSTER_DRAIN_TIMEOUT)
-        # Set the duration to an hour, if we haven't cleaned up and termintated by then
-        # mesos should put the slave back into the pool
-        duration = 600
-        log.info("Draining {0}".format(instances_to_kill))
-        log.info("Decreasing spot fleet capacity to: {0}".format(target_capacity))
-        if not dry_run:
-            # sort to make testing easier
-            drain([instance['ip'] for instance in sorted(instances_to_kill.values())], start, duration)
-            ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
-                                                 ExcessCapacityTerminationPolicy='noTermination')
-        log.info("Waiting for instances to drain before we terminate")
-        wait_and_terminate(instances_to_kill, dry_run)
+        sfr_sorted_slaves.reverse()
+        while True:
+            if len(sfr_sorted_slaves) == 0:
+                break
+            slave_to_kill = sfr_sorted_slaves.pop()
```

Comment:
Agreed 👍 

Ive added some more checks but kept the `pop()` method. I think that it makes more sense because if something goes wrong we shouldn't retry the slave or add it back to the list (we could end up in an infinite loop etc.) And the whole thing gets retried in the next run anyway.